rust/libnewsboat/src/utils.rs

```diff
@@ -268,32 +259,22 @@ pub fn is_valid_attribute(attribute: &str) -> bool {
     VALID_ATTRIBUTES.contains(&attribute)
 }
 
-pub fn strwidth(rs_str: &str) -> usize {
-    let control = rs_str.chars().fold(true, |acc, x| acc & !x.is_control());
-
-    if control {
-        return UnicodeWidthStr::width(rs_str);
+pub fn strwidth(s: &str) -> usize {
+    if s.chars().any(|x| !x.is_control()) {
```

Comment:
I think this should be `all()`, not `any()`:
1. on an empty string, original `fold()` would return `true`. `any()` would return `false`, while `all()` would return `true` ⇒ we should use `all()`;
2. on a string with a mix of ordinary and control characters, original `fold()` would return `false` (because at some point `!x.is_control()` will return `false`, and `true & false = false`; the `&` is an error I guess, should be `&&` really). `any()` would return `true` (because there is an ordinary character for which the predicate returns `true`), while `all()` would return `false` (because there is a control character for which the predicate returns `false`) ⇒ we should use `all()`.

For strings consisting entirely of ordinary *or* entirely of control characters, all three variants return the same values (`true` and `false`, respectively).

Please revert to `fold`, add two `assert`s to check empty and mixed strings (and copy them to C++ tests as well), make sure they pass, then bring back your changes (with `all()` this time). No need to do it in separate commits, but do follow these steps to make sure the tests do pass for the previous version of the code.

Reply:
I do not get what you meant to say. Why not just `UnicodeWidth::width(s)` for everything? Then there is no need for extra iteration to check for control characters.

Reply:
Oh, that's a great question! As it turns out, the control characters check [was added because wcswidth returns -1 if it finds such character](https://github.com/newsboat/newsboat/commit/3c1ea127c553ba28904aa15816a2d9bd115ed9dd). `UnicodeWidth::width(&str)` [treats control chars as zero-width](https://unicode-rs.github.io/unicode-width/unicode_width/trait.UnicodeWidthStr.html#tymethod.width), so we indeed can replace strwidth with a call to that.

(What I meant to say originally is: this new code doesn't behave the same as the old one. It should use `all()` instead of `any()`. But that's not important now.)

Reply:
I tested but it does not seemed to work. `assert_eq!(strwidth_stfl("\u{0007}"), 1);` fail since that returns 0.

Reply:
Note that \u{0007} is a control character ([a bell](https://en.wikipedia.org/wiki/Bell_character)). This test checks the workaround which we now remove. The workaround counts bytes rather than characters, thus overestimating the length. `UnicodeWidth` will return the exact count, so that test should be changed from `1` to `0`.

Please link to https://github.com/newsboat/newsboat/pull/518#discussion_r283100228 from the commit message, so the next person running into this has some context.

Reply:
Okay, updated this on the latest commit.

Reply:
Not quite fixed it yet—the C++ tests weren't updated; please apply the same changes you made to Rust ones.

Reply:
Should be fixed.

Reply:
Ah, you updated one but missed the other one: https://travis-ci.org/newsboat/newsboat/jobs/532132534#L1269

Reply:
Ops, sorry about that. Fixed.